### PR TITLE
feat: Make num-bigint optional

### LINKED ISSUE #124: Would you consider making bigint support optional?
`num-bigint` is not small — I'd like to make bigint support optional in [merde_json](https://github.com/bearcove/merde_json) and I'd be happy to submit a PR to put bigint support behind a feature flag in jiter.

### PR DESCRIPTION
**Pull request by fasterthanlime**
Closes #124 

Apologies for the few TOML reformats that slipped through, I can get rid of those if needed.

A few statics got turned into consts and moved inline with the BigInt code to avoid "unused variable" warnings.

I would recommend running `cargo hack --feature-powerset check` in CI, from https://lib.rs/crates/cargo-hack

---

A "default features build" (with num-bigint) is 4.7s at best for me:

CleanShot 2024-08-16 at 17 40 21@2x

Command used:

```shell
cargo clean ; cargo build --package jiter --quiet --timings && w3m -dump target/*timings/*timing.html | grep 'Total time.*$' --color=always -B3
```

And a "no default features" build is just under two seconds:

CleanShot 2024-08-16 at 17 41 36@2x

Command used:

```shell
cargo clean ; cargo build --package jiter --quiet --timings --no-default-features && w3m -dump target/*timings/*timing.html | grep 'Total time.*$' --color=always -B3
```

### CODE REVIEW

**Review by samuelcolvin — APPROVED** [review]
LGTM, thanks so much.

@davidhewitt you happy with this?

I tried to merge with main to fix the fuzz failures, but you've blocked contributions from maintainers, @fasterthanlime could you rebase please?

**Comment by fasterthanlime** [comment]
> @fasterthanlime could you rebase please?

Done!

**Comment by fasterthanlime** [comment]
(Is there a way to "Allow edits by maintainers" after the fact?)

**Comment by samuelcolvin** [comment]
Thanks so much.

> (Is there a way to "Allow edits by maintainers" after the fact?)

I think there might be a way, perhaps a checkbox once if you edit the PR body?

Anyway, happy to merge this once @davidhewitt has a quick check.

**Review by davidhewitt — APPROVED** [review]
Super, thanks very much!

**Comment by fasterthanlime** [comment]
> Super, thanks very much!

Thanks for the review+merge! Can you tag me when this lands on crates.io?

(By the way I recommend [the excellent release-plz](https://release-plz.ieni.dev) for release automation, in case you're interested)

**Comment by cetra3** [comment]
I've raised a followup PR for this one as it looks like `num-bigint` is always pulling in pyo3 as well: https://github.com/pydantic/jiter/pull/160

**Pull request merged**
